### DISCUSSION TITLE
Basic infos for compiling

### DISCUSSION THREAD

**Discussion by MasterPlexus** [j-kaltes/Juggluco#15] [Q&A]
At first, many thanks for the work!

Anyhow I have some questions, because up to now I was not able to build Juggluco by my own. So I would like to ask the following things:

- in the linker are some folders shown, which are not clear to me, for example "-I/n/ojka/src/android/frida/build/sdk-android-arm64/include". This looks like a include folder for headerfiles, but SDK I'm not sure to. So to what this folder belongs to? Is may /usr/include/ the right one as also openssl and so on is needed?
- Wich Android-Studio version, and wich build tools are you using for build (jvm 11 or lower, which ndk)?
- in cpp/net/watchserver are two files (servercrt.h and serverkey.h) which are sym-linked to the folder /n/ojka/prog/ssl/mx.juggluco.nl/. I assume those needs to be get from a other source?
- in Typeinfo I get several cast errors, is there may something wrong? But I assume this is belonging may wrong cc version...
```
cast from pointer to smaller type 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') loses information
reinterpret_cast from 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') to 'size_t' (aka 'unsigned long') is not allowed
cast from pointer to smaller type 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') loses information
reinterpret_cast from 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') to 'size_t' (aka 'unsigned long') is not allowed
cast from pointer to smaller type 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') loses information
reinterpret_cast from 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') to 'size_t' (aka 'unsigned long') is not allowed
cast from pointer to smaller type 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') loses information
reinterpret_cast from 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') to 'size_t' (aka 'unsigned long') is not allowed
cast from pointer to smaller type 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') loses information
reinterpret_cast from 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') to 'size_t' (aka 'unsigned long') is not allowed
cast from pointer to smaller type 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') loses information
reinterpret_cast from 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') to 'size_t' (aka 'unsigned long') is not allowed
```
I assume, that I'm may just need some hints to get further, but up to now I'm not clear on it.
BR
MP

**Comment by j-kaltes** [comment]
You don't need servercrt.h and serverkey.h
I nearly always use the latest of everything (except that I don't use preview versions):
You can find them in the Common/build.gradle file on this github.
grep ndkVersion build.gradle 
    ndkVersion '25.2.9519653'
grep compileSdk build.gradle
    compileSdk 33
grep buildToolsVersion build.gradle 
    buildToolsVersion '33.0.0'
    buildToolsVersion '33.0.2'

/home/jka/android-studio/jbr/bin/java --version
openjdk 11.0.15 2022-04-19
OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
OpenJDK 64-Bit Server VM (build 11.0.15+0-b2043.56-8887301, mixed mode)

When I remove /n/ojka/src/android/frida/build/sdk-android-arm64/include, I get the following error message:
[...] cpp/net/watchserver/sslserver.cpp:36:10: fatal error: 'openssl/ssl.h' file not found
#include "openssl/ssl.h"

**Reply by MasterPlexus** [reply]
> When I remove /n/ojka/src/android/frida/build/sdk-android-arm64/include, I get the following error message:
> [...] cpp/net/watchserver/sslserver.cpp:36:10: fatal error: 'openssl/ssl.h' file not found
> #include "openssl/ssl.h"
So then it looks like your standart includes of aditive packages are stored there... I made a symlink of that now to my /usr/include, and openssl is then found (i do it that way, because I had problems to include the standard path in the build).

Anyhow it looks like still the build brings up some problems from typeinfo (wich is part of the ndk package). Really curious, as this is a standard referenced to ndk (it refers to 
```
In file included from /home/dirk/StudioProjects/Juggluco/Common/src/main/cpp/net/libreview/putsensor.cpp:25:
In file included from /home/dirk/StudioProjects/Juggluco/Common/src/main/cpp/settings/settings.h:23:
In file included from /home/dirk/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/array:110:
In file included from /home/dirk/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/iterator:421:
In file included from /home/dirk/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/__functional_base:15:
/home/dirk/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/typeinfo:216:14: error: cast from pointer to smaller type 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') loses information
```
)
I'm now back on newest versions, similar as you explained (thanks for that). Maybe there is a wrong order in the header search and a other typeinfo is searched?

By the way, which build task is the current correct one for the normal libre 3 apk (I use mobileLibre3release)

**Comment by j-kaltes** [comment]
mobileLibre3release is the Libre3 version for phones.
You are now talking about problems in a system header file. Do you use Macintosh? Above it I read "This implementation is specific to ARM64 on Apple platforms.". Juggluco is compiled with -fno-rtti , so maybe you can remove it.

**Comment by MasterPlexus** [comment]
I'm on Lubuntu. I will give it a try,with your latest suggestion. thanks

**Comment by MasterPlexus** [comment]
So now I tried several things. The proposed switch from you did not change something. Anyhow I could see the mentioned errors are not the only ones related to ndk type info. Also some errors in netstuff.h happens 

```
no member named 'in6_u' in 'in6_addr'
no member named 'in6_u' in 'in6_addr'
field designator 'in6_u' does not refer to any field in type 'const struct in6_addr'
field designator 'in6_u' does not refer to any field in type 'const struct in6_addr'

```
or in libreview.h

```
In file included from /home/dirk/StudioProjects/Juggluco/Common/src/main/cpp/net/libreview/newlibre3.cpp:36:
/home/dirk/StudioProjects/Juggluco/Common/src/main/cpp/net/libreview/libreview.h:40:23: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
```
And there will be may more, as the build then stops.

Im not sure, but for me it looks like that there are may some dependency's in your build system wich are not fullfilled on my side. As I stated I'm on lubuntu. You are on Mac? may that's the reason, but currently I have no further idea on the possible differency's. Is here may someone which has no problems under Linux, where I could compair with?

**Reply by j-kaltes** [reply]
I also use Linux. Either you have changed something in the source of Juggluco or something in your Android studio installation is different.  Did you find out why earlier you were compiling code above which stood "This implementation is specific to ARM64 on Apple platforms."?

**Reply by MasterPlexus** [reply]
This message I never get, where did you see that?

Of course interims I did some changes, but nothing which touches those files I assume. Anyhow I will try a complete new download and retry...

**Reply by j-kaltes** [reply]
Sorry, I make a mistake.
This remark is at line 154 of Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/typeinfo, but when I read everything I saw that this is a very general description of different implementations. It didn't mean that everything in that part of the file was an Apple implementation. When I compiled I also used that part of the file.
You problem has to be something else.

**Comment by MasterPlexus** [comment]
So now restart with a new download. the only thing I adopted is the sym link between /usr/include and /n/ojka/src/android/frida/build/sdk-android-arm64/include

Then after build, the first error I've get is 
```
/home/dirk/StudioProjects/Juggluco_new/Common/src/main/cpp/net/watchserver/watchserver.cpp:48:10: fatal error: 'sha1.hpp' file not found
/home/dirk/StudioProjects/Juggluco_new/Common/src/main/cpp/net/watchserver/watchserver.cpp:48:10: fatal error: 'sha1.hpp' file not found
/home/dirk/StudioProjects/Juggluco_new/Common/src/main/cpp/net/watchserver/watchserver.cpp:48:10: fatal error: 'sha1.hpp' file not found
/home/dirk/StudioProjects/Juggluco_new/Common/src/main/cpp/net/watchserver/watchserver.cpp:48:10: fatal error: 'sha1.hpp' file not found
```
Anyhow, the file is availibel in the area:
```
/n/ojka/src/android/frida/build/sdk-android-arm64/include/boost/beast/core/detail/sha1.hpp
/n/ojka/src/android/frida/build/sdk-android-arm64/include/boost/compute/detail/sha1.hpp
/n/ojka/src/android/frida/build/sdk-android-arm64/include/boost/uuid/detail/sha1.hpp
```
In the build before I solved that because included my /usr/include in cmake all stages. may this was not the right way.

Could you explain, where you have placed which headers coming from Linux installation in your build machine?

Anyway thanks for your try's to help ;)

**Comment by j-kaltes** [comment]
I forgot to upload sha1.hpp. I have now done so. I have also put the openssl header files in the source of Juggluco. Just clone the Juggluco github again.

**Comment by MasterPlexus** [comment]
The header files of openssl i did not see in the source of last commit. but anyhow, I guess this is not needed, as it will come from standart linux headers (and could make you need to update every time if a vulnerablility of openssl will be committed).

anyhow, now it stuck at this point:
```
Natives.java
    package com.eveningoutpost.dexdrip.services.broadcastservice.models does not exist
    cannot find symbol class GraphLine
    cannot find symbol class GraphLine
    cannot find symbol class GraphLine
WearInt.java
    package com.eveningoutpost.dexdrip.services.broadcastservice.models does not exist
    package com.eveningoutpost.dexdrip.services.broadcastservice.models does not exist
    cannot find symbol class Settings
    cannot find symbol class Settings
    cannot find symbol class Settings
watchdrip.java
    package com.eveningoutpost.dexdrip.services.broadcastservice.models does not exist
```

**Reply by j-kaltes** [reply]
The Openssl header files I had already uploaded earlier. I have just now also uploaded GraphLine.

**Reply by MasterPlexus** [reply]
now I got a finished build, thanks!!

Now I can start with the next steps ;)

**Comment by MasterPlexus** [comment]
missed headers and explanations are given, and build is now working. Thanks.
